src/nnet3/nnet-chain-training.cc

```diff
@@ -147,6 +149,44 @@ void NnetChainTrainer::ProcessOutputs(const NnetChainExample &eg,
   }
 }
 
+void NnetChainTrainer::DoScalingFactorLearning() {
+  const NnetTrainerOptions &nnet_config = opts_.nnet_config;
+  if (nnet_config.scale_learning_rate <= 0.0 &&
+      RandUniform() > nnet_config.scale_update_frequency)
+    return;
+  // Note: if we are using momentum, the magnitude of the parameter changes in
+  // delta_nnet_ will be larger by a factor of 1.0 / (1.0 - nnet_config.momentum),
+  // which comes from the sum of 1, momentum, momentum^2,...; the factor of
+  // (1.0 - nnet_config.momentum) is to compensate for this. The factor of
+  // 1.0 / nnet_config.scale_update_frequency compensates for the fact that we
+  // don't do this on all minibatches.
+  BaseFloat scale_learning_rate = nnet_config.scale_learning_rate *
+      (1.0 - nnet_config.momentum) * (1.0 / nnet_config.scale_update_frequency);
+
+  int32 i = 0;
+  for (int32 c = 0; c < delta_nnet_->NumComponents(); c++) {
+    Component *comp = delta_nnet_->GetComponent(c);
+    if (comp->Properties() & kUpdatableComponent) {
+      UpdatableComponent *uc_delta = dynamic_cast<UpdatableComponent*>(comp);
+      UpdatableComponent *uc_nnet =
+        dynamic_cast<UpdatableComponent*>(nnet_->GetComponent(c));
+      if (uc_delta == NULL || uc_nnet == NULL)
+        KALDI_ERR << "Updatable component does not inherit from class "
+                  << "UpdatableComponent; change this code.";
+      // computes the dot product of the parameter in nnet_ with the parameter
+      // in delta_nnet_, which is the derivatives w.r.t. the scaling factor 
+      BaseFloat dot_prod =uc_delta->DotProduct(*uc_nnet);
+      // adds the param changes due to the change of parameter's scaling factor,
+      // to nnet_delta_.
+      BaseFloat alpha = scale_learning_rate * dot_prod;
+      uc_delta->Add(alpha, *uc_nnet);
```

Comment:
You'll have to be a bit careful here about the interaction with momentum.  This formula would not work correctly if momentum != 0.0; it would diverge, because you're adding to delta_nnet and next time you work out the scaling factor from delta_nnet.  One way to fix this would be to remember the 'alpha' values and subtract them again after adding delta_nnet to nnet, after calling  AddNnetComponents().